rules/Php85/Rector/Class_/FinalPropertyPromotionRector.php

```diff
@@ -0,0 +1,175 @@
+    }
+
+    /**
+     * @param Class_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $scope = ScopeFetcher::fetch($node);
+        if ($this->shouldSkip($node, $scope)) {
+            return null;
+        }
+
+        $constructClassMethod = $node->getMethod(MethodName::CONSTRUCT);
+
+        if (! $constructClassMethod instanceof ClassMethod) {
+            return null;
+        }
+
+        foreach ($constructClassMethod->getParams() as $param) {
+            $phpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($param);
+
+            if (! $phpDocInfo->hasByName(self::TAGNAME)) {
+                continue;
+            }
+            $this->visibilityManipulator->makeFinal($param);
+            $phpDocInfo->removeByName(self::TAGNAME);
+            $this->docBlockUpdater->updateRefactoredNodeWithPhpDocInfo($param);
+        }
+
+        return $node;
+    }
+
+    public function provideMinPhpVersion(): int
+    {
+        return PhpVersionFeature::FINAL_PROPERTY_PROMOTION;
+    }
+
+    /**
+     * @return ClassReflection[]
+     */
+    private function resolveParentClassReflections(Scope $scope): array
+    {
+        $classReflection = $scope->getClassReflection();
+        if (! $classReflection instanceof ClassReflection) {
+            return [];
+        }
+
+        return $classReflection->getParents();
+    }
+
+    private function shouldSkip(Class_ $class, Scope $scope): bool
+    {
+        $classReflection = $scope->getClassReflection();
+        if (! $classReflection instanceof ClassReflection) {
+            return true;
+        }
+
+        if ($this->shouldSkipClass($class)) {
+            return true;
+        }
+
+        $parents = $this->resolveParentClassReflections($scope);
+
+        if ($class->isFinal()) {
+            return true;
+        }
+
+        $constructClassMethod = $class->getMethod(MethodName::CONSTRUCT);
+        if (! $constructClassMethod instanceof ClassMethod) {
+            return false;
+        }
+        $params = $constructClassMethod->getParams();
+
+        return $this->shouldSkipParams($params);
+    }
+
+    private function shouldSkipClass(Class_ $class): bool
+    {
+        if ($this->visibilityManipulator->hasVisibility($class, Visibility::FINAL)) {
+            return true;
+        }
+
+        if ($this->phpAttributeAnalyzer->hasPhpAttribute($class, AttributeName::ALLOW_DYNAMIC_PROPERTIES)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @param Param[] $params
+     */
+    private function shouldSkipParams(array $params): bool
+    {
+        foreach ($params as $param) {
+            if ($this->visibilityManipulator->hasVisibility($param, Visibility::FINAL) && $param->isPromoted()) {
+                return true;
+            }
+
+            // type is missing, invalid syntax
+            if ($param->type === null) {
+                return true;
+            }
```

Comment:
non-typed is allowed for final promoted property https://3v4l.org/mVo8m/rfc#vgit.master